File: spsc-lite-haskell/HE.hs
module HE(he, embeddedIn) where

import SLanguage
import Algebra

-- Enhanced homeomorphic embedding:
-- expressions are compared only if they belong
-- to the same category (as defined by `aVarIsUnderAttack`).

embeddedIn e1 e2 =
  aVarIsUnderAttack e1 == aVarIsUnderAttack e2 && he e1 e2 

-- This is the "classic" homeomorphic imbedding relation.

he e1 e2 = heByDiving e1 e2 || heByCoupling e1 e2

heByDiving e1 (Var _) = False
heByDiving e1 (Call _ _ args) = any (he e1) args

heByCoupling (Var _) (Var _) = True
heByCoupling (Call kind1 name1 args1) (Call kind2 name2 args2)
               | kind1 == kind2 && name1 == name2 =
      and (zipWith he args1 args2)
heByCoupling _ _ = False

-- We distinguish a specific category of expressions:
-- the ones that generate contractions in the process tree.

aVarIsUnderAttack (Call GCall _ args) = aVarIsUnderAttack (head args)
aVarIsUnderAttack (Var _) = True
aVarIsUnderAttack _ = False

File: spsc-lite-scala-paper-v1/src/spsc/HE.scala
package spsc

import Algebra._

object HE {
  def he_*(t1: Term, t2: Term): Boolean = he(t1, t2) && b(t1) == b(t2)
  
  def he(t1: Term, t2: Term) = heByDiving(t1, t2) || heByCoupling(t1, t2)
  
  private def heByDiving(t1: Term, t2: Term): Boolean = t2 match {
    case e: CFG => e.args exists (he(t1, _))
    case _ => false
  }
  
  private def heByCoupling(t1: Term, t2: Term): Boolean = (t1, t2) match {
    case (e1:CFG, e2:CFG) if shallowEq(e1, e2) => List.forall2(e1.args, e2.args)(he)
    case (Var(_), Var(_)) => true
    case _ => false
  }
  
  private def b(t: Term): Int = t match {
    case GCall(_, args) => b(args.head)
    case Var(_) => 1
    case _ => 0
  } 
}